src/github.com/matrix-org/go-neb/services/jira/jira.go

```diff
@@ -0,0 +1,168 @@
+						return nil, errors.New("Failed to map project key to a JIRA endpoint.")
+					}
+					if r == nil {
+						return nil, errors.New("No known project exists with that project key.")
+					}
+
+					iss := jira.Issue{
+						Fields: &jira.IssueFields{
+							Summary:     title,
+							Description: desc,
+							Project: jira.Project{
+								Key: pkey,
+							},
+							// FIXME: This may vary depending on the JIRA install!
+							Type: jira.IssueType{
+								Name: "Bug",
+							},
+						},
+					}
+					cli, err := r.JIRAClient(userID, false)
+					i, res, err := cli.Issue.Create(&iss)
+					if err != nil {
+						log.WithError(err).Print("Failed to create issue")
+						return nil, errors.New("Failed to create issue")
+					}
+					if res.StatusCode < 200 || res.StatusCode >= 300 {
+						return nil, fmt.Errorf("Failed to create issue: JIRA returned %d", res.StatusCode)
+					}
+
+					return &matrix.TextMessage{
+						"m.notice",
+						fmt.Sprintf("Created issue: %s", i.Key),
+					}, nil
+				},
+			},
+		},
+	}
+}
+func (s *jiraService) OnReceiveWebhook(w http.ResponseWriter, req *http.Request, cli *matrix.Client) {
+	w.WriteHeader(200) // Do nothing
+}
+
+func (s *jiraService) projectToRealm(userID, pkey string) (*realms.JIRARealm, error) {
+	// We don't know which JIRA installation this project maps to, so:
+	//  - Get all known JIRA realms and f.e query their endpoints with the
+	//    given user ID's credentials (so if it is a private project they
+	//    can see it will succeed.)
+	//  - If there is a matching project with that key, return that realm.
+	// We search installations which the user has already OAuthed with first as most likely
+	// the project key will be on a JIRA they have access to.
+	// TODO: Return whether they have authed or not so they know if they need to make a starter link
+	logger := log.WithFields(log.Fields{
+		"user_id": userID,
+		"project": pkey,
+	})
+	knownRealms, err := database.GetServiceDB().LoadAuthRealmsByType("jira")
+	if err != nil {
+		logger.WithError(err).Print("Failed to load jira auth realms")
+		return nil, err
+	}
+	// typecast and move ones which the user has authed with to the front of the queue
+	var queue []*realms.JIRARealm
+	for _, r := range knownRealms {
+		jrealm, ok := r.(*realms.JIRARealm)
+		if !ok {
+			logger.WithField("realm_id", r.ID()).Print(
+				"Failed to type-cast 'jira' type realm into JIRARealm")
+			continue
+		}
+
+		_, err := database.GetServiceDB().LoadAuthSessionByUser(r.ID(), userID)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				// not authed with this JIRA realm; add to back of queue.
+				queue = append(queue, jrealm)
+			} else {
+				logger.WithError(err).WithField("realm_id", r.ID()).Print(
+					"Failed to load auth sessions for user")
+			}
+			continue // this may not have been the match anyway so don't give up!
+		}
+		// authed with this JIRA realm; add to front of queue
+		queue = append([]*realms.JIRARealm{jrealm}, queue...)
```

Comment:
Would it make sense to keep two queues here and to merge them together at the end?
That way the realms will be added in the same order for authed and non authed realms.